src/DirectoryListing/index.ts

```diff
@@ -0,0 +1,162 @@
+   * return the next location for list or null if the recursion should stop and yield the current item.
+   * For advanced usage you can also limit the depth of recursion using the second argument of next function.
+   */
+  public recursive(
+    next: (current: T, depth: number, driver: Driver) => Promise<string | null> | string | null = (
+      current
+    ) => (current.isFile ? null : current.location)
+  ) {
+    // get copy of current chain until call to recursive
+    // so we can apply it to listing when doing recursion
+    const chain = this.chain.slice()
+
+    return this.pipe(async function* (source) {
+      for await (const item of source) {
+        // call the next function to know if we should continue recursion
+        const location = await next(item, 0, this)
+
+        if (location === null) {
+          yield item
+        } else {
+          // create next function to pass to recursive for next level of recursion
+          // we are using the original function if it is not using the depth parameter
+          const nextWithDepth =
+            next.length > 1
+              ? (current: T, depth: number, driver: Driver) => next(current, depth + 1, driver)
+              : next
+
+          // list the returned location from next function and also apply the chain to it
+          const list = chain
+            .reduce((listing, fn) => listing.pipe(fn), this.list(location))
+            .recursive(nextWithDepth)
+
+          const iterator = list[Symbol.asyncIterator]()
+          let nextItem = await iterator.next()
+
+          // if the directory is empty it means it is a leaf and we will yield it
+          // else we will yield the items in the directory
+          if (nextItem.done) {
+            yield item
+          } else {
+            do {
+              yield nextItem.value
+              nextItem = await iterator.next()
+            } while (!nextItem.done)
+          }
+        }
+      }
+    })
+  }
+
+  /**
+   * Add a piping chain function which gets the current async iterable and returns
+   * new async iterable with modified directory listing output.
+   * Function this is bound to instance of driver for which the listing is generated.
+   * This allows using async generator functions and reference the driver methods easily.
+   * Piping will always return clone of the current instance and add the function
+   * to the chain of new cloned instance only to prevent side effects.
+   */
+  public pipe<U>(fn: (this: Driver, iterable: AsyncIterable<T>) => AsyncIterable<U>) {
+    const clone = new DirectoryListing(this.driver, this.listing)
+    clone.chain = this.chain.concat(fn)
+    return clone as unknown as DirectoryListingContract<Driver, U>
+  }
+
+  /**
+   * Get the final async iterable after passing directory listing through chain of piping functions modifying the output.
+   */
+  public toIterable(): AsyncIterable<T> {
```

Comment:
Any reason for not implementing this as `Symbol.asyncIterator` ?

Reply:
Do you mean why I have created a function and not use this code directly in `Symbol.asyncIterator` below ? Mainly for reusing it in different methods like `toArray` (maybe others when extending). I can change it if it is a problem.

Reply:
Sorry, I stopped scrolling and commented when I saw that `toArray` called this method instead of doing `for await (const item of this)`, assuming `Symbol.asyncIterator` wasn't implemented at all.